### PR TITLE
introduce new integration: dyn-traffic-director

### PR DESCRIPTION
**Pull request by jfchevrette**
Ref.: https://issues.redhat.com/browse/APPSRE-3884

Depends on schema changes: https://github.com/app-sre/qontract-schemas/pull/7

Notes for reviewers:
- The Dyn module is old and has not been updated in almost 4 years. It works well for our purpose and avoids having to re-implement their REST API. The Dyn service is being retired so we should not expect changes in their APIs or the module: https://www.oracle.com/corporate/acquisitions/dyn/technologies/enterprise-customer-faq.html
- I've used type hints where it made sense
- Unfortunately no tests. I think it may be acceptable to not have tests considering this is a service that is being retired and not being updated. I'm happy to add tests if deemed absolutely necessary

### CODE REVIEW

File: reconcile/dyn_traffic_director.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                                raise UnsupportedRecordType(
+                                    f'record type {type(record)} is not '
+                                    f'supported: {record}'
+                                )
+
+                            cname_rdata = record.rdata()['cname_rdata']
+                            if 'cname' not in cname_rdata:
+                                raise UnsupportedRecordType(
+                                    f'record missing a cname field: {record}'
+                                )
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py:70]: The integration supports only CNAME records. We're not going to manage existing TD configs

The idea is if it ever comes across a record type it can't understand it will error out and we can intervene

**Comment by jfchevrette** [comment]
> How soon is this service getting retired? Are we on the May 2022 path or on the 2023 one? Depending on which one it is, tests will be more or less important.

Late 2023 according to the FAQ

File: reconcile/dyn_traffic_director.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        td_ttl: int = td.get('ttl', DEFAULT_TD_TTL)
+
+        records: List[Dict[str, Union[str, int]]] = []
+        for record in td_records:
+            if record['cluster']:
+                hostname = record['cluster']['elbFQDN']
+            elif record['hostname']:
+                hostname = record['hostname']
+            else:
+                raise InvalidRecord('either cluster or hostname must '
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py:112]: Added more info in the message in ac2173ff

File: reconcile/dyn_traffic_director.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+class InvalidRecord(Exception):
+    pass
+
+
+class UnsupportedRecordType(Exception):
+    pass
+
+
+def fetch_current_state() -> dict:
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: fixed in f4531231

[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: fixed in f4531231

File: reconcile/dyn_traffic_director.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    state: dict = {
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py:40]: Thanks for the suggestion! I've fixed and improved the return type hint in f4531231

File: reconcile/dyn_traffic_director.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                            records.append({
+                                'hostname':
+                                # strip trailing dot added by Dyn
+                                cname_rdata['cname'].rstrip('.'),
+                                'weight':
+                                int(cname_rdata.get('weight', DEFAULT_WEIGHT)),
+                            })
+
+        # need to be sorted for comparison later
+        records = sorted(records, key=lambda d: d['hostname'])
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Thanks for the suggestion! Changed in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/20bd7443d2fa1c4fa398067c29d4c5e13d54dbaa

File: reconcile/dyn_traffic_director.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+        }
+
+    return state
+
+
+def process_tds(current: dict, desired: dict,
+                dry_run: bool = True, enable_deletion: bool = False) -> bool:
+    errors = False
+
+    added: List[str] = [name for name in desired if name not in current]
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Implemented suggestion in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/8d91aa7d17c49fdaab99fd812cc1f7d239c72e45

File: reconcile/dyn_traffic_director.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        for dname, d in desired.items()
+        for cname, c in current.items()
+        if dname == cname if d != c
+    ]
+
+    # Process removed TD services
+    for name in removed:
+        td_name = current[name]['name']
+        msg = ['delete_td']
+        msg.append(name)
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Implemented this suggestions in all occurrences in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/89849e524e019ac90ef49af0c24a1bbfe6e7c7ab

File: reconcile/dyn_traffic_director.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        msg = ['create_td']
+        msg.append(name)
+        logging.info(msg)
+
+        records = []
+        for record in td_records:
+            records.append(
+                DSFCNAMERecord(record['hostname'], label='',
+                               automation='manual', weight=record['weight'])
+            )
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: I don't think it makes the code much clearer but I've implemented it anyway :)
https://github.com/app-sre/qontract-reconcile/pull/1965/commits/9f1063f3cbd0f1e33d3bde48a9c212ed7ae06723

File: reconcile/dyn_traffic_director.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+                curtd.delete()
+
+    # Process added TD services
+    for name in added:
+        td_name = desired[name]['name']
+        td_ttl = desired[name]['ttl']
+        td_records = desired[name]['records']
+
+        msg = ['create_td']
+        msg.append(name)
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: implemented suggestion in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/89849e524e019ac90ef49af0c24a1bbfe6e7c7ab

File: reconcile/dyn_traffic_director.py [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        dname
+        for dname, d in desired.items()
+        for cname, c in current.items()
+        if dname == cname if d != c
+    ]
+
+    # Process removed TD services
+    for name in removed:
+        td_name = current[name]['name']
+        msg = ['delete_td']
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: implemented suggestion in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/89849e524e019ac90ef49af0c24a1bbfe6e7c7ab

File: reconcile/dyn_traffic_director.py [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+        records = []
+        for record in td_records:
+            records.append(
+                DSFCNAMERecord(record['hostname'], label='',
+                               automation='manual', weight=record['weight'])
+            )
+
+        attach_node = None
+        zones = dyn_zones.get_all_zones()
+        for zone in zones:
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: I agree! https://github.com/app-sre/qontract-reconcile/pull/1965/commits/4dccea09cf410cda1df59cb92707a34785827316

File: reconcile/dyn_traffic_director.py [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+        record_set = DSFRecordSet('CNAME', label=td_name,
+                                  automation='manual', records=records)
+        failover_chain = DSFFailoverChain(label=td_name,
+                                          record_sets=[record_set])
+        rpool = DSFResponsePool(label=td_name, rs_chains=[failover_chain])
+        ruleset = DSFRuleset(label=td_name, criteria_type='always',
+                             response_pools=[rpool])
+        TrafficDirector(td_name, ttl=td_ttl, rulesets=[ruleset],
+                        nodes=[attach_node])
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Yeah I can't agree more with your guys... Those APIs/Objects were a bit of headache to use. The alternative was to re-implement their REST API which IMO isn't worth the extra work for the little value we would get out of it. That service is not being updated anymore and is being retired in 2 years.

I've added a comment in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/9203ed0c387a78fd1f081efb09c6ad7c56d079a4

File: reconcile/dyn_traffic_director.py [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        update_td: TrafficDirector = None
+        for td in dyn_services.get_all_dsf_services():
+            if td.label == name:
+                update_td = td
+
+        # Check if ttl need to be updated
+        if current[name]['ttl'] != desired[name]['ttl']:
+            msg = ['update_td_ttl']
+            msg.append(name)
+            msg.append(desired[name]['ttl'])
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: implemented in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/89849e524e019ac90ef49af0c24a1bbfe6e7c7ab

File: reconcile/dyn_traffic_director.py [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+
+            if not dry_run:
+                # Generate a new list of CNAME records
+                records = []
+                for record in desired[name]['records']:
+                    records.append(
+                        DSFCNAMERecord(record['hostname'], label='',
+                                       automation='manual',
+                                       weight=record['weight'])
+                    )
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: refactored in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/61b7e638d1c0f5ff23d9c9a5e773261bb79e724e

**Comment by jfchevrette** [comment]
> Providing some feedback on type hint side. Thanks for adding these, it definitely makes the code easier to follow. I have less background on the Dyn side, but out of curiosity do we have a general date for when this integration would be deprecated?

Late 2023 according to the FAQ

File: reconcile/dyn_traffic_director.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Thanks for the tip! I have added more specific type hints in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/8af318cc0e3b8b85beaf8fb2efa0be97ba7ac6df

File: reconcile/dyn_traffic_director.py [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+            'name': td_name,
+            'nodes': [td_name],
+            'ttl': td_ttl,
+            'records': records,
+        }
+
+    return state
+
+
+def process_tds(current: dict, desired: dict,
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: I have added more specific type hints in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/8af318cc0e3b8b85beaf8fb2efa0be97ba7ac6df

I think this is further against what you are suggesting here? I'd like to better understand why this can be an issue but here unless you think we really should not be doing this I would go ahead and leave it as is...

[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: I agree that it's long but IMO the 3 for loops /w comments make it easy to understand. Would you be against keeping it as-is?

[the remark this answers is not in the document]
Comment (steveteahan) [reply reconcile/dyn_traffic_director.py]: I think there were two separate issues:

1. The other comment where I pointed out that using `Any` implicitly, like in `Dict` where that is really `Dict[Any, Any]` leads to needing to provide other types later in the method because there simply weren't hints to know that keys were strings.
2. Using abstract types for parameters is better because it makes methods more general.

Putting the two together we'd go from something like:

```
def my_method(data: Dict)
```
to

```
# Can use MutableMapping for cases where the object is being modified
def my_method(data: Mapping[str, Any]
```

That's just an example, but now the method can accept any Mapping (including a dictionary) and it has specific enough types that we know that the keys are `str`s, which might save us from providing other collection types explicitly. Sorry if providing this feedback separately made it seem contradicting (combined with it `Iterable` not having a type in one of the examples).

[the remark this answers is not in the document]
Comment (steveteahan) [reply reconcile/dyn_traffic_director.py]: Fixed style guide to hopefully avoid this confusion in the future: https://gitlab.cee.redhat.com/service/dev-guidelines/-/merge_requests/90

[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: Improved in https://github.com/app-sre/qontract-reconcile/pull/1965/commits/1be499840fbdfedee8791e1932ba45fd1fc9fe40
Thanks for the additional explanation!

[the remark this answers is not in the document]
Comment (Piojo) [reply reconcile/dyn_traffic_director.py]: Long methods are very hard to test, and I'd really appreciate not reducing coverage in this repo. Jenkins is coplaining that this brings coverage to less than 26.45%, our current minimum.

If you can add tests (we are going to maintain this for almost two years) and address the type suggestions, I can leave it as it is.

[the remark this answers is not in the document]
Comment (maorfr) [reply reconcile/dyn_traffic_director.py]: a "long" method is somewhat objective. the method upholds everything written here: https://gitlab.cee.redhat.com/service/dev-guidelines/-/blob/master/content/en/docs/AppSRE/python-style-guide.md#limit-method-size

**Review by maorfr — APPROVED** [review]

**Comment by jfchevrette** [comment]
I've done a fairly big refactor to address concerns around long methods (ie: process_tds) and concerns around low test coverage

@Piojo @steveteahan could I request another review? TIA!

File: reconcile/dyn_traffic_director.py [hunk 16]
```diff
@@ -0,0 +1,511 @@
...
+            if not attach_node:
+                raise UpdateTrafficDirectorError(
+                    f'Could not find DNS node {name}')
+            td.nodes = [attach_node]
+
+
+def process_tds(current: Mapping[str, Mapping[str, Any]],
+                desired: Mapping[str, Mapping[str, Any]],
+                dry_run: bool = True,
+                enable_deletion: bool = False):
```
Comment (Piojo) [inline reconcile/dyn_traffic_director.py:458]:
```suggestion
                enable_deletion: bool = False) -> None:
```

**Review by Piojo — APPROVED** [review]
Very small comments, no real blockers IMO.

File: reconcile/test/test_dyn_traffic_director.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+    desired = generate_state(2, 1, 3, 30)
+
+    mock_create_td = mocker.patch.object(integ, 'create_td', autospec=True)
+    mock_delete_td = mocker.patch.object(integ, 'delete_td', autospec=True)
+    mock_update_td = mocker.patch.object(integ, 'update_td', autospec=True)
+
+    integ.process_tds(current, desired, dry_run=True,
+                      enable_deletion=False)
+
+    assert mock_create_td.called
```
Comment (steveteahan) [inline reconcile/test/test_dyn_traffic_director.py:99]: I'd either do `assert_called_once()` or to ensure that it's the record that is expected, `assert_called_once_with` which allows checking the parameters that were passed in: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with

File: reconcile/test/test_dyn_traffic_director.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+                      enable_deletion=False)
+
+    assert not mock_create_td.called
+    assert not mock_delete_td.called
+    assert not mock_update_td.called
+
+
+def test_process_tds_added_td(mocker):
+    """Tests that TDs are added given an additional TD in desired state"""
+    current = generate_state(1, 1, 3, 30)
```
Comment (steveteahan) [inline reconcile/test/test_dyn_traffic_director.py:89]: Sometimes it's also helpful to test the edge-cases like empty state. I'll leave that up to you as far as whether that can happen or not when adding new resources.

**Review by steveteahan — COMMENTED** [review]
Thanks for adding these tests! They're helpful in checking that we'll only call create/delete/update when expected. The dryrun switch is within those methods, so that is one critical thing that might be worth testing. It seems like that wouldn't be the most difficult test because you can just make sure that the helper methods that are doing the API calls aren't called (or the object itself). I'll leave it up to you and others who have a better idea of the deadline here if we want to take the risk of that part being untested.

File: reconcile/test/test_dyn_traffic_director.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    mock_create_td = mocker.patch.object(integ, 'create_td', autospec=True)
+    mock_delete_td = mocker.patch.object(integ, 'delete_td', autospec=True)
+    mock_update_td = mocker.patch.object(integ, 'update_td', autospec=True)
+
+    integ.process_tds(current, desired, dry_run=True,
+                      enable_deletion=False)
+
+    assert not mock_create_td.called
+    assert mock_delete_td.called
+    assert not mock_update_td.called
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/test/test_dyn_traffic_director.py]: fixed in 96d97121

File: reconcile/test/test_dyn_traffic_director.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+    mock_create_td = mocker.patch.object(integ, 'create_td', autospec=True)
+    mock_delete_td = mocker.patch.object(integ, 'delete_td', autospec=True)
+    mock_update_td = mocker.patch.object(integ, 'update_td', autospec=True)
+
+    integ.process_tds(current, desired, dry_run=True,
+                      enable_deletion=False)
+
+    assert not mock_create_td.called
+    assert not mock_delete_td.called
+    assert mock_update_td.called
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/test/test_dyn_traffic_director.py]: fixed in 96d97121

File: reconcile/test/test_dyn_traffic_director.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+    desired = generate_state(1, 1, 3, 30)
+
+    mock_create_td = mocker.patch.object(integ, 'create_td', autospec=True)
+    mock_delete_td = mocker.patch.object(integ, 'delete_td', autospec=True)
+    mock_update_td = mocker.patch.object(integ, 'update_td', autospec=True)
+
+    integ.process_tds(current, desired, dry_run=True,
+                      enable_deletion=False)
+
+    assert not mock_create_td.called
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/test/test_dyn_traffic_director.py]: fixed in 96d97121

File: reconcile/test/test_dyn_traffic_director.py (hunk 17 above)
Comment (jfchevrette) [reply reconcile/test/test_dyn_traffic_director.py]: fixed in 96d97121

File: reconcile/dyn_traffic_director.py [hunk 22]
```diff
@@ -0,0 +1,511 @@
...
+        ruleset = _get_dyn_traffic_director_ruleset(td, ruleset_label)
+        rpool = _get_dyn_traffic_director_response_pool(ruleset, rpool_label)
+        chain = _get_dyn_traffic_director_chain(rpool, chain_label)
+        recordset = _get_dyn_traffic_director_recordset(chain, rset_label)
+    except DynResourceNotFound as e:
+        raise DynResourceNotFound(
+            f'could not retrieve records for traffic director: {td.label}, '
+            f'ruleset: {ruleset_label}, resourcepool: {rpool_label}, '
+            f'failoverchain: {chain_label}, recordset: {rset_label}. '
+            f'The specific error was: {e}'
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: fixed in 5be82bbb

File: reconcile/dyn_traffic_director.py [hunk 23]
```diff
@@ -0,0 +1,511 @@
...
+        chain = _get_dyn_traffic_director_chain(rpool, chain_label)
+        current_recordset = _get_dyn_traffic_director_recordset(chain,
+                                                                rset_label)
+    except DynResourceNotFound as e:
+        raise DynResourceNotFound(
+            f'could not retrieve records for traffic director: {td.label}, '
+            f'ruleset: {ruleset_label}, resourcepool: {rpool_label}, '
+            f'failoverchain: {chain_label}, recordset: {rset_label}. '
+            f'The specific error was: {e}'
+        )
```
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: fixed in 5be82bbb

File: reconcile/dyn_traffic_director.py (hunk 16 above)
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: added return type hint in 3363ac6c

File: reconcile/test/test_dyn_traffic_director.py (hunk 18 above)
Comment (jfchevrette) [reply reconcile/test/test_dyn_traffic_director.py:106]: I've added a test in 854939a5

File: reconcile/dyn_traffic_director.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jfchevrette) [reply reconcile/dyn_traffic_director.py]: the method has been split and shortened quite a bit
